README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

Spline chart library for Flutter. Spline chart is a line chart that uses curves instead of straight lines.

## Features

![Screenshot 1](https://raw.githubusercontent.com/medicaljoyworks/spline_chart/main/doc/images/screenshot1.png)
![Screenshot 2](https://raw.githubusercontent.com/medicaljoyworks/spline_chart/main/doc/images/screenshot2.png)
![Screenshot 3](https://raw.githubusercontent.com/medicaljoyworks/spline_chart/main/doc/images/screenshot3.png)

## Usage

Import the Spline Chart package and use the SplineChart widget.

```dart
import 'package:spline_chart/spline_chart.dart';

SplineChart(
  values: {0:10,15:30,50:10,44:100,80:50,100:90},
  verticalLineEnabled: true,
  verticalLinePosition: 90.0,
  verticalLineStrokeWidth: 2.0,
  verticalLineText: "The middle",
  drawCircles: true,
  circleRadius: 5,
  width: 320,
  height: 400,
)
```

## Additional information

This plugin is developed by [Medical Joyworks](https://www.medicaljoyworks.com/).

CHANGELOG.md
## 0.0.2

Minor changes to follow Dart file conventions and fixing static analysis issues.

## 0.0.1

The initial release of the library.


test/spline_chart_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:spline_chart/spline_chart.dart';

void main() {
  test('TBD', () {
    //final chart = SplineChart();
  });
}


lib/spline_chart.dart
library spline_chart;

import 'dart:math';

import 'package:flutter/material.dart';
import 'package:intl/intl.dart' as intl;

/// A Spline chart widget.
///
/// This widget draws a spline chart based on the given [values].
class SplineChart extends StatelessWidget {
  /// Width of the chart
  final double width;

  /// Height of the chart
  final double height;

  /// Line chart values
  ///
  /// Each key in the hash map represents the X position and it's value as the Y position
  final Map<double, double> values;

  /// Start of the X axis
  ///
  /// Defaults to 0
  final double xStart;

  /// End of the X axis
  ///
  /// Defaults to 100
  final double xEnd;

  /// The X axis lable intervals
  ///
  /// Defaults to 10
  final double xStep;

  /// Color of the line being drawn
  final Color lineColor;

  /// Color of the grid lines
  final Color gridLineColor;

  /// Color of the label texts
  final Color textColor;

  /// Size of the label texts
  final double textSize;

  /// Thickness of the line
  final double strokeWidth;

  /// Color of the fill
  final Color fillColor;

  /// Opacity of the fill
  final double fillOpactiy;

  /// If true, a vertical line will be drawn at [verticalLinePosition]
  final bool verticalLineEnabled;

  /// Position of the vertical line
  final double verticalLinePosition;

  /// Thickness of the vertical line
  final double verticalLineStrokeWidth;

  /// Color of the vertical line
  final Color verticalLineColor;

  /// The text for the vertical line label
  ///
  /// Leave it empty if no label is needed
  final String? verticalLineText;

  /// If true, each data point would be highlighted by circles
  final bool drawCircles;

  /// Fill color of the data point circles
  final Color circleFillColor;

  /// Thickness of the data point circles
  final Color circleStrokeColor;

  /// Radius of the data point circles
  final double circleRadius;

  const SplineChart(
      {Key? key,
      required this.values,
      this.width = 320.0,
      this.height = 200.0,
      this.lineColor = Colors.black,
      this.gridLineColor = Colors.grey,
      this.textColor = Colors.grey,
      this.textSize = 14,
      this.xStart = 0,
      this.xEnd = 100,
      this.xStep = 10,
      this.strokeWidth = 1,
      this.fillColor = Colors.lightBlue,
      this.fillOpactiy = 0.5,
      this.verticalLineEnabled = false,
      this.verticalLinePosition = 0.0,
      this.verticalLineStrokeWidth = 1.0,
      this.verticalLineColor = Colors.red,
      this.verticalLineText,
      this.drawCircles = false,
      this.circleFillColor = Colors.lightBlueAccent,
      this.circleStrokeColor = Colors.black,
      this.circleRadius = 5})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      child: SizedBox(
        width: this.width,
        height: this.height,
        child: CustomPaint(
            size: Size(this.width, this.height),
            painter: _SplineChartPainter(
              lineColor: this.lineColor,
              gridLineColor: this.gridLineColor,
              textColor: this.textColor,
              textSize: this.textSize,
              values: this.values,
              xStart: this.xStart,
              xEnd: this.xEnd,
              xStep: this.xStep,
              strokeWidth: this.strokeWidth,
              fillColor: this.fillColor,
              fillOpactiy: this.fillOpactiy,
              verticalLineColor: this.verticalLineColor,
              verticalLineEnabled: this.verticalLineEnabled,
              verticalLinePosition: this.verticalLinePosition,
              verticalLineStrokeWidth: this.verticalLineStrokeWidth,
              verticalLineText: this.verticalLineText,
              drawCircles: this.drawCircles,
              circleFillColor: this.circleFillColor,
              circleStrokeColor: this.circleStrokeColor,
              circleRadius: this.circleRadius,
            )),
      ),
    );
  }
}

class _SplineChartPainter extends CustomPainter {
  final Color lineColor;
  final Color gridLineColor;
  final Color textColor;
  final double textSize;
  final Color fillColor;
  final double fillOpactiy;
  final Map<double, double> values;
  final xStart;
  final xEnd;
  final xStep;
  final double strokeWidth;
  final bool verticalLineEnabled;
  final double verticalLinePosition;
  final double verticalLineStrokeWidth;
  final Color verticalLineColor;
  final String? verticalLineText;
  final bool drawCircles;
  final double circleRadius;
  final Color circleFillColor;
  final Color circleStrokeColor;

  _SplineChartPainter({
    required this.lineColor,
    required this.gridLineColor,
    required this.textColor,
    required this.textSize,
    required this.values,
    required this.xStart,
    required this.xEnd,
    required this.xStep,
    required this.strokeWidth,
    required this.fillColor,
    required this.fillOpactiy,
    required this.verticalLineEnabled,
    required this.verticalLinePosition,
    required this.verticalLineStrokeWidth,
    required this.verticalLineColor,
    this.verticalLineText,
    required this.drawCircles,
    required this.circleStrokeColor,
    required this.circleFillColor,
    required this.circleRadius,
  });

  double _calcStepSize(double range, int targetSteps) {
    double tempStep = range / targetSteps;
    var magPow = pow(10, (log(tempStep) / ln10).floor());
    var magMsd = (tempStep / magPow + 0.5);
    if (magMsd > 5)
      magMsd = 10;
    else if (magMsd > 2)
      magMsd = 5;
    else if (magMsd > 1) magMsd = 2;
    return magMsd * magPow;
  }

  @override
  void paint(Canvas canvas, Size size) {
    final paint = Paint()
      ..color = lineColor
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth;

    double yMin = double.infinity, yMax = double.negativeInfinity;
    values.forEach((key, value) {
      yMin = min(value, yMin);
      yMax = max(value, yMax);
    });

    double yStep = _calcStepSize(yMax - yMin, 20);
    yMin = yMin - (yMin % yStep);
    yMax = ((yMax / yStep).floor() + 1) * yStep;

    // draw grid lines
    final axisLinePaint = Paint()
      ..color = gridLineColor
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth + strokeWidth / 2;

    final gridLinePaint = Paint()
      ..color = gridLineColor
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth - strokeWidth / 2;

    double paddingX = 50;
    double paddingY = 40;

    // yAxis
    canvas.drawLine(Offset(paddingX, 0),
        Offset(paddingX, size.height - paddingY), axisLinePaint);
    // xAxis
    canvas.drawLine(Offset(paddingX, size.height - paddingY),
        Offset(size.width, size.height - paddingY), axisLinePaint);

    intl.NumberFormat numberFormat = intl.NumberFormat.decimalPattern('hi');

    // draw vertical grid lines
    double xRatio = (size.width - paddingX) / (this.xEnd - this.xStart);
    for (double x = this.xStart; x <= this.xEnd; x += this.xStep) {
      canvas.drawLine(Offset(x * xRatio + paddingX, 0),
          Offset(x * xRatio + paddingX, size.height - paddingY), gridLinePaint);

      TextSpan span = TextSpan(
          style: TextStyle(color: this.textColor, fontSize: this.textSize),
          text: numberFormat.format(x.floor()));
      TextPainter tp =
          TextPainter(text: span, textDirection: TextDirection.ltr);
      tp.layout();
      tp.paint(
          canvas,
          Offset(x * xRatio + paddingX - tp.width / 2,
              size.height - paddingY + 5));
    }

    // draw horizontal grid lines
    double yRatio = (size.height - paddingY) / (yMax - yMin);
    for (double y = yMin; y <= (yMax); y += yStep) {
      double yPos = (size.height - paddingY) - ((y - yMin) * yRatio);
      canvas.drawLine(
          Offset(paddingX, yPos), Offset(size.width, yPos), gridLinePaint);

      TextSpan span = TextSpan(
          style: TextStyle(color: this.textColor, fontSize: this.textSize),
          text: numberFormat.format(y.floor()));
      TextPainter tp =
          TextPainter(text: span, textDirection: TextDirection.ltr);
      tp.layout();
      tp.paint(canvas, Offset(paddingX - tp.width - 5, yPos - tp.height / 2));
    }

    // sort values by x position
    List<double> xValues = this.values.keys.toList();
    xValues.sort((a, b) => (a - b).floor());

    List<Offset> circles = [];

    final path = Path();
    for (int i = 0; i < xValues.length; i++) {
      if (i == 0) {
        if (xValues[0] == 0) {
          path.moveTo(
              paddingX,
              size.height -
                  paddingY -
                  ((values[xValues[0]] ?? 0) - yMin) * yRatio);
          circles.add(Offset(
              paddingX,
              size.height -
                  paddingY -
                  ((values[xValues[0]] ?? 0) - yMin) * yRatio));
        } else {
          path.moveTo(paddingX, size.height - paddingY - yMin);
          circles.add(Offset(paddingX, size.height - paddingY - yMin));
        }
      } else {
        final yPrevious = size.height -
            paddingY -
            ((values[xValues[i - 1]] ?? 0) - yMin) * yRatio;
        final xPrevious = xValues[i - 1] * xRatio + paddingX;
        final controlPointX =
            xPrevious + (xValues[i] * xRatio + paddingX - xPrevious) / 2;

        final yValue = size.height -
            paddingY -
            ((values[xValues[i]] ?? 0) - yMin) * yRatio;

        path.cubicTo(controlPointX, yPrevious, controlPointX, yValue,
            xValues[i] * xRatio + paddingX, yValue);

        circles.add(Offset(xValues[i] * xRatio + paddingX, yValue));
      }
    }
    canvas.drawPath(path, paint);

    if (drawCircles) {
      final circleFillPaint = Paint()
        ..color = circleFillColor
        ..style = PaintingStyle.fill
        ..strokeWidth = strokeWidth;

      final circleStrokePaint = Paint()
        ..color = circleStrokeColor
        ..style = PaintingStyle.stroke
        ..strokeWidth = strokeWidth;

      circles.forEach((offset) {
        canvas.drawCircle(offset, circleRadius, circleFillPaint);
        canvas.drawCircle(offset, circleRadius, circleStrokePaint);
      });
    }

    Path fillPath = Path()..addPath(path, Offset.zero);
    fillPath.relativeLineTo(strokeWidth / 2, 0.0);
    fillPath.lineTo(size.width, size.height - paddingY);
    fillPath.lineTo(size.width + strokeWidth / 2, size.height - paddingY);
    fillPath.lineTo(paddingX, size.height - paddingY);
    fillPath.close();

    Paint fillPaint = Paint()
      ..strokeWidth = 0.0
      ..color = fillColor.withOpacity(fillOpactiy)
      ..style = PaintingStyle.fill;
    canvas.drawPath(fillPath, fillPaint);

    if (verticalLineEnabled) {
      final verticalLinePaint = Paint()
        ..color = verticalLineColor
        ..style = PaintingStyle.stroke
        ..strokeWidth = verticalLineStrokeWidth;

      canvas.drawLine(
          Offset(verticalLinePosition * xRatio + paddingX, 0),
          Offset(
              verticalLinePosition * xRatio + paddingX, size.height - paddingY),
          verticalLinePaint);

      if (this.verticalLineText != null) {
        TextSpan span = TextSpan(
          style: TextStyle(color: verticalLineColor, fontSize: this.textSize),
          text: this.verticalLineText,
        );
        TextPainter tp = TextPainter(
          text: span,
          textDirection: TextDirection.ltr,
        );
        tp.layout();
        double x = verticalLinePosition * xRatio + paddingX + 5;
        if (x + tp.width > size.width) {
          x = verticalLinePosition * xRatio + paddingX - tp.width - 5;
        }
        tp.paint(canvas, Offset(x, 5));
      }
    }
  }

  @override
  bool shouldRepaint(_SplineChartPainter old) {
    return old.lineColor != lineColor ||
        old.gridLineColor != gridLineColor ||
        old.textColor != textColor ||
        old.textSize != textSize ||
        old.fillColor != fillColor ||
        old.fillOpactiy != fillOpactiy ||
        old.values != values ||
        old.xStart != xStart ||
        old.xEnd != xEnd ||
        old.xStep != xStep ||
        old.strokeWidth != strokeWidth ||
        old.verticalLineEnabled != verticalLineEnabled ||
        old.verticalLinePosition != verticalLinePosition ||
        old.verticalLineStrokeWidth != verticalLineStrokeWidth ||
        old.verticalLineColor != verticalLineColor ||
        old.verticalLineText != verticalLineText ||
        old.drawCircles != drawCircles ||
        old.circleRadius != circleRadius ||
        old.circleFillColor != circleFillColor ||
        old.circleStrokeColor != circleStrokeColor;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:spline_chart/spline_chart.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const SplineChartDemo(),
    );
  }
}

class SplineChartDemo extends StatelessWidget {
  const SplineChartDemo({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.white,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceAround,
        children: [
          SplineChart(
            values: {0:10,15:30,60:40,44:50,80:50,100:90},
            //values: {0:10, 7.69 :52, 15.38 :74, 23.08 :1464, 30.77 :942, 38.46 :2433, 46.15 :2379, 53.85 :3820, 61.54 :2750, 69.23 :2739, 76.92 :3057, 84.62 :1598, 92.31 :1450,100:630},
            //values: {0.0: 583.0, 50.0: 972.0, 100.0: 910.0},
            verticalLineEnabled: false,
            verticalLinePosition: 90.0,
            verticalLineStrokeWidth: 2.0,
            verticalLineText: "Your score",
            drawCircles: true,
            circleRadius: 4,
            width: 320,
            height: 400,
          ),
        ],
      ),
    );
  }
}